datalad_next/iter_collections/gitworktree.py

```diff
@@ -250,23 +255,27 @@ def _lsfiles_line2props(
 
 
 def _git_ls_files(path, *args):
-    # we use a plain runner to avoid the overhead of a GitRepo instance
-    runner = ThreadedRunner(
-        cmd=[
-            'git', 'ls-files',
-            # we rely on zero-byte splitting below
-            '-z',
-            # otherwise take whatever is coming in
-            *args,
-        ],
-        protocol_class=StdOutCaptureGeneratorProtocol,
-        stdin=DEVNULL,
-        # run in the directory we want info on
-        cwd=path,
-    )
-    line_splitter = LineSplitter('\0', keep_ends=False)
-    # for each command output chunk received by the runner
-    for content in runner.run():
-        # for each zerobyte-delimited "line" in the output
-        for line in line_splitter.process(content.decode('utf-8')):
-            yield line
+    with run(
+            cmd=[
+                'git', 'ls-files',
+                # we rely on zero-byte splitting below
+                '-z',
+                # otherwise take whatever is coming in
+                *args,
+            ],
+            protocol_class=StdOutCaptureGeneratorProtocol,
+            input=DEVNULL,
+            cwd=path
+    ) as r:
+        # This code uses the data processor chain to process data. This fixes
+        # a problem with the previous version of the code, where `decode` was
+        # used on every data chunk that was sent tp `pipe_data_received`. But
+        # data is chunked up randomly and might be split in the middle of a
+        # character encoding, leading to weird errors.
```

Comment:
This comments make me think whether could be any legitimate use of `run()` without the decoder construct below, whenever output is to be processed?

I think not.

This raises the questions if anything useful/typical could/would be done with `r` (context manager yield), before it is consumed by `process_from()`? If there isn't something, then this would suggest that spiltting these components up may make sense internally, but is needless complexity for a user to deal with. `processors=` could become an argument of `run()`.

Reply:
> This raises the questions if anything useful/typical could/would be done with `r` (context manager yield), before it is consumed by `process_from()`?

`r` holds the exit code of the process, so it makes sense for the user to keep a copy around. But I like the idea of `processor=`.  We could copy the exit-code to the yielded  `process_from()`-generator, when the __exit__-method is executed. So the user-interface would actually not change.

Reply:
> This comments make me think whether could be any legitimate use of run() without the decoder construct below, whenever output is to be processed?

Output processing does not necessarily include decoding. A remote ssh cat-command would want to receive bytes and still require processing, for example, to look for a specific pattern in the returned bytes.